onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "qorder_attention.h"
+#include "contrib_ops/cuda/bert/attention_impl.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/shared_inc/fpgeneric.h"
+
+using namespace onnxruntime::cuda;
+using namespace ::onnxruntime::common;
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+#define REGISTER_KERNEL_TYPED(T)                                  \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
+      QOrderedAttention,                                          \
+      kMSDomain,                                                  \
+      1,                                                          \
+      T,                                                          \
+      kCudaExecutionProvider,                                     \
+      (*KernelDefBuilder::Create())                               \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      QOrderedAttention);
+
+REGISTER_KERNEL_TYPED(int8_t)
+
+QOrderedAttention::QOrderedAttention(const OpKernelInfo& info) : CudaKernel(info), AttentionBase(info) {
+  order_input_ = GetCublasLtOrderAttr(info, "order_input");
+  order_weight_ = GetCublasLtOrderAttr(info, "order_weight");
+  order_bias_ = GetCublasLtOrderAttr(info, "order_bias");
+  order_output_ = GetCublasLtOrderAttr(info, "order_output");
+  //TODO: any checks?
+  // bugbug
+}
+
+Status QOrderedAttention::ComputeInternal(OpKernelContext* context) const {
+  // inputs are column based
+  const Tensor* input = context->Input<Tensor>(0);
+  const Tensor* weights = context->Input<Tensor>(2);
+  const Tensor* bias = context->Input<Tensor>(4);
+  const Tensor* mask_index = context->Input<Tensor>(6);
+
+  auto& device_prop = GetDeviceProp();
+  ORT_RETURN_IF_ERROR(CheckInputs(input->Shape(), weights->Shape(), bias->Shape(), mask_index, nullptr, nullptr, device_prop.maxThreadsPerBlock));
+
+  const Tensor* scale_input = context->Input<Tensor>(1);
+  const Tensor* scale_weights = context->Input<Tensor>(3);
+  const Tensor* scale_bias = context->Input<Tensor>(5);
+  const Tensor* scale_output = context->Input<Tensor>(7);
+
+  const float* scale_input_data = scale_input->template Data<float>();
+  const float* scale_weights_data = scale_weights->template Data<float>();
+  const float* scale_bias_data = scale_bias->template Data<float>();
+  const float* scale_output_data = scale_output->template Data<float>();
+
+  // bugbug
+  ORT_UNUSED_PARAMETER(scale_output_data);
```

Comment:
question: not sure what's the last scale in this op and how to use it

Reply:
You could just scale_bias_data = scale_bias_data / scale_output_data;